##################### 转换
library(reticulate)
library(Seurat)
library(patchwork)
library(stringr)
library(Matrix)
library(sceasy)
library(dior)
library(schard)
use_python("G:/miniconda3/python.exe")

####################################
library(biomaRt)
library(SummarizedExperiment)
library(TCGAbiolinks)

check_package <- function(package){
    if (!requireNamespace(package, quietly = TRUE)) {
        stop(package, " package is needed for this function to work. Please install it.",
             call. = FALSE)
    }
}

get.GRCh.bioMart <- function(
    genome = c("hg19", "hg38"),
    as.granges = FALSE
) {

    genome <- match.arg(genome)
    # Since the amount of users complaining about the access we
    # also added the data into a data package
    check_package("TCGAbiolinksGUI.data")
    if (genome == "hg19") {
        gene.location <- get(
            data("gene.location.hg19",
                 package = "TCGAbiolinksGUI.data",
                 envir = environment())
        )

        if (as.granges) {
            gene.location$strand[gene.location$strand == 1] <- "+"
            gene.location$strand[gene.location$strand == -1] <- "-"
            gene.location$chromosome_name <- paste0("chr",gene.location$chromosome_name)
            gene.location <- makeGRangesFromDataFrame(
                gene.location, seqnames.field = "chromosome_name",
                start.field = "start_position",
                end.field = "end_position",
                keep.extra.columns = TRUE
            )
        }
    } else {
        gene.location <- get(
            data(
                "gencode.v36.annotation.genes",
                package = "TCGAbiolinksGUI.data",
                envir = environment()
            )
        )
        if(!as.granges) gene.location <- as.data.frame(gene.location)
    }

    return(gene.location)
}

map.ensg <- function(genome = "hg38", genes) {
    gene.location <- get.GRCh.bioMart(genome)
    gene.location <- gene.location[match(genes, gsub("\\.[0-9]*","",gene.location$gene_id)), ]
    return(gene.location)
}

gene.location <- get.GRCh.bioMart("hg38")
gene.location <- gene.location[, c("gene_name","gene_type")]
gene.location <- gene.location[gene.location$gene_type == 'protein_coding',]
#########################
setwd('H:\\CKDwork\\knokdown\\Data')
h5ad_file <- './Mature_Full_v3.h5ad'

sc <- schard::h5ad2seurat(h5ad_file,use.raw=F)
sc <- sc[rownames(sc) %in% gene.location$gene_name,]

setwd('H:\\CKDwork\\knokdown\\Data')
saveRDS(sc, file = 'Mature_Full_v3.rds')

setwd('H:\\CKDwork\\knokdown\\Data')
h5ad_file <- './Fetal_full_v3.h5ad'

sc <- schard::h5ad2seurat(h5ad_file,use.raw=F)
sc <- sc[rownames(sc) %in% gene.location$gene_name,]

setwd('H:\\CKDwork\\knokdown\\Data')
saveRDS(sc, file = 'Fetal_full_v3.rds')

